### ISSUE TITLE
thread 'main' panicked at 'already borrowed: BorrowMutError'

### ISSUE THREAD

**Issue by mikhasd** [CraneStation/wasmtime#176]
I have just built wasmtime and tried to run a simple command:

`~/dev/wasmtime/target/release/wasmtime graph.wasm`

```
thread 'main' panicked at 'already borrowed: BorrowMutError', src/libcore/result.rs:999:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:212
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:475
   7: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:382
   8: rust_begin_unwind
             at src/libstd/panicking.rs:309
   9: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  10: core::result::unwrap_failed
  11: wasmtime_init_finish
  12: wasmtime_runtime::instance::InstanceHandle::new
  13: wasmtime_wast::spectest::instantiate_spectest
  14: wasmtime::main
  15: std::rt::lang_start::{{closure}}
  16: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  17: std::panicking::try::do_call
             at src/libstd/panicking.rs:294
  18: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:82
  19: std::panicking::try
             at src/libstd/panicking.rs:273
  20: std::panic::catch_unwind
             at src/libstd/panic.rs:388
  21: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  22: main
  23: __libc_start_main
  24: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:347:21
stack backtrace:
   0:     0x55595b8c2fcb - backtrace::backtrace::libunwind::trace::hee16fdad5997e164
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1:     0x55595b8c2fcb - backtrace::backtrace::trace_unsynchronized::haf0450f72372f592
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2:     0x55595b8c2fcb - std::sys_common::backtrace::_print::h05ce790c31111ac9
                               at src/libstd/sys_common/backtrace.rs:47
   3:     0x55595b8c2fcb - std::sys_common::backtrace::print::h0ff493cc4d2d2ae6
                               at src/libstd/sys_common/backtrace.rs:36
   4:     0x55595b8c2fcb - std::panicking::default_hook::{{closure}}::hab4f96382fb4b194
                               at src/libstd/panicking.rs:198
   5:     0x55595b8c2cce - std::panicking::default_hook::h8edc7382c3bc85e5
                               at src/libstd/panicking.rs:212
   6:     0x55595b8c3740 - std::panicking::rust_panic_with_hook::hcd9f879ec989ce4d
                               at src/libstd/panicking.rs:475
   7:     0x55595b8c32c2 - std::panicking::continue_panic_fmt::h4b92ca51a5d8f90c
                               at src/libstd/panicking.rs:382
   8:     0x55595b8c31a6 - rust_begin_unwind
                               at src/libstd/panicking.rs:309
   9:     0x55595b8dcffd - core::panicking::panic_fmt::h76420a1d3d63d38a
                               at src/libcore/panicking.rs:85
  10:     0x55595b8dcf3c - core::panicking::panic::h7988f6b8a74ce9b4
                               at src/libcore/panicking.rs:49
  11:     0x55595b64e1da - std::thread::local::LocalKey<T>::with::h407f07886df8e514
  12:     0x55595b64aafd - Unwind
  13:     0x55595b650db3 - _ZL15WasmTrapHandleriP9siginfo_tPv
  14:     0x7ff5bd216f40 - <unknown>
  15:     0x55595b649b62 - wasmtime_init_finish
  16:     0x55595b648126 - wasmtime_runtime::instance::InstanceHandle::new::h21223c5f21299e02
  17:     0x55595b5c2d92 - wasmtime_wast::spectest::instantiate_spectest::h51b0a58177bf3199
  18:     0x55595b597dd6 - wasmtime::main::ha9be6962cd6baa7b
  19:     0x55595b595fd3 - std::rt::lang_start::{{closure}}::hc7b0fd2866a9683c
  20:     0x55595b8c3143 - std::rt::lang_start_internal::{{closure}}::h0d9d3a9c3cc0a33e
                               at src/libstd/rt.rs:49
  21:     0x55595b8c3143 - std::panicking::try::do_call::h8a3e545414bd3807
                               at src/libstd/panicking.rs:294
  22:     0x55595b8c66aa - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:82
  23:     0x55595b8c3d0d - std::panicking::try::h9e35c35221d74c26
                               at src/libstd/panicking.rs:273
  24:     0x55595b8c3d0d - std::panic::catch_unwind::h30f9df8c252fe03c
                               at src/libstd/panic.rs:388
  25:     0x55595b8c3d0d - std::rt::lang_start_internal::h075e7b80ec0300cc
                               at src/libstd/rt.rs:48
  26:     0x55595b59a9e2 - main
  27:     0x7ff5bd024b6b - __libc_start_main
  28:     0x55595b594bfa - _start
  29:                0x0 - <unknown>
thread panicked while panicking. aborting.
thread panicked while processing panic. aborting.
thread panicked while processing panic. aborting.
```

I have used rust nightly to compile it:
```
rustc 1.37.0-nightly (2fe7b3383 2019-06-19
```

**Comment by sunfishcode** [comment]
Thanks for the report! Which platform are you seeing this on? Also, are you able to post the graph.wasm file somewhere?

**Comment by mikhasd** [comment]
Hi! you can find the binary [here](https://send.firefox.com/download/73747d142aaaeb51/#PMqfW8owIQAETLfOWIrQ3g) and the source code [here](https://github.com/mikhasd/wasm)

**Comment by bbigras** [comment]
I might have an old wasmtime (20190521).

[wasi-hello-world.zip](https://github.com/CraneStation/wasmtime/files/3785933/wasi-hello-world.zip)

```
RUST_BACKTRACE=full wasmtime target/wasm32-wasi/debug/wasi-hello-world.wasm
thread 'main' panicked at 'already borrowed: BorrowMutError', src/libcore/result.rs:1084:5
stack backtrace:
   0:     0x5567dd55b771 - std::sys_common::backtrace::print::h74e8383011c48ac1
   1:     0x5567dd550e14 - std::panicking::default_hook::{{closure}}::hc989d29eac1bf07a
   2:     0x5567dd550ba3 - std::panicking::default_hook::h999474d3b9874310
   3:     0x5567dd551520 - std::panicking::rust_panic_with_hook::h8ac69ca134e1594e
   4:     0x5567dd5510a2 - std::panicking::continue_panic_fmt::ha458e61c9ee535d1
   5:     0x5567dd550f86 - rust_begin_unwind
   6:     0x5567dd575f0d - core::panicking::panic_fmt::hdbb1e8356e4e7099
   7:     0x5567dd56fa27 - core::result::unwrap_failed::h0e98f81ef171a4f8
   8:     0x5567dd2f6379 - wasmtime_init_finish
   9:     0x5567dd2f49c6 - wasmtime_runtime::instance::InstanceHandle::new::h12eeccb065d5f844
  10:     0x5567dd2588cb - wasmtime_wast::spectest::instantiate_spectest::hfcc9d5dd0cd83c76
  11:     0x5567dd22ee46 - wasmtime::main::he8880607d4290727
  12:     0x5567dd22cfd3 - std::rt::lang_start::{{closure}}::h6d689296becf02cb
  13:     0x5567dd550f23 - std::panicking::try::do_call::h0c14233a80d9596c
  14:     0x5567dd55d47a - __rust_maybe_catch_panic
  15:     0x5567dd553749 - std::rt::lang_start_internal::h34c921f5951345b4
  16:     0x5567dd231ad5 - main
  17:     0x7f7850882b8e - __libc_start_main
  18:     0x5567dd22bbfa - _start
  19:                0x0 - <unknown>
```

**Comment by peterhuene** [comment]
Hi @bbigras.  It does look like your wasmtime version is older than what I believe to be the fixed version.  I am able to run your wasm file without issue using both the latest out of master and the `0.3.0` release.

If you [upgrade your wasmtime](https://github.com/CraneStation/wasmtime/releases/tag/v0.3.0) to `0.3.0`, that should help fix the issue.

As I can not reproduce this failure any longer and I'm fairly confident the fix was in #253, I believe we can close this issue.  Please let me know if a newer version of wasmtime does not fix your issue.  Thanks!

**Comment by bbigras** [comment]
Sorry for the late reply. NixOS unstable now has 20191018 and using it fixed my issue. Thanks.
